### PR TITLE
fix(ls): windows fixes

### PR DESCRIPTION
**Pull request by ajbt200128**
This PR includes some small fixes for LSP.js in its build system. It also fixes some common functions that didn't properly respect windows line endings (\r\n).

## Test plan
copy js/language_server/dist to semgrep-vscode's lspjs folder and run it on a windows machine :)

See https://github.com/semgrep/semgrep-vscode/pull/80 for tests

### CODE REVIEW

**Review by tpetr — APPROVED** [review]
LGTM but might want a 👍 from a PA person for stuff outside of the js folder

File: src/osemgrep/language_server/scan_helpers/Scan_helpers.ml [hunk 1]
```diff
@@ -41,6 +41,9 @@ let run_semgrep ?(targets = None) ?(rules = None) ?(git_ref = None)
   if rules = [] then (
     Logs.debug (fun m -> m "No rules to run! Not scanning anything.");
     ([], []))
+  else if targets = Some [] then (
+    Logs.debug (fun m -> m "No targets to scan! Not scanning anything.");
+    ([], []))
```
Comment (kopecs) [inline src/osemgrep/language_server/scan_helpers/Scan_helpers.ml]:
```
match rules, targets with
  [], _ -> (* stuff *)
  _, Some [] -> (* stuff *)
  _ -> (* stuff *)
```
nicer?

File: languages/cpp/menhir/flag_parsing_cpp.ml [hunk 2]
```diff
@@ -8,7 +8,14 @@ type language = C | Cplusplus
...
 (*****************************************************************************)
 
-let macros_h = ref (Fpath.v "/data/cpp_stdlib/macros.h")
+let macros_h =
+  ref
+    (Fpath.v
+       (match Sys.win32 with
+       (* Windows doesn't like paths that start with / *)
+       (* So provide a default path that may exist in windows *)
+       | true -> "C:/Program Files/LLVM/include/c++/v1/macros.h"
```
Comment (kopecs) [inline languages/cpp/menhir/flag_parsing_cpp.ml:17]: Would be nice to see this justified. As is I don't understand what purpose this file path serves, so I think more docs would highly improve quality.

File: js/shared/core.js [hunk 3]
```diff
@@ -171,8 +188,21 @@ function unix_open(path, flags, perm) {
...
 
+  let file;
   /* inlined from caml_sys_open */
-  var file = root.device.open(root.rest, f);
+
+  // Node is weird and doesn't like opening NUL on Windows
+  // if its fs.openSync("NUL", "rs+") call, so we have to
+  // special case it and use \\\\.\\nul instead.
+  if (globalThis.process.platform === "win32" && path == "NUL") {
+    path = "\\\\.\\nul";
```
Comment (kopecs) [inline js/shared/core.js:198]:
```suggestion
    path = {|\\.\nul|};
```
Think you can do this instead?

Comment (kopecs) [reply js/shared/core.js:198]: Sorry. This is js, not ocaml. Could do `` String.raw`<string>` ``.

File: js/shared/core.js [hunk 4]
```diff
@@ -171,8 +188,21 @@ function unix_open(path, flags, perm) {
...
   }
 
+  let file;
   /* inlined from caml_sys_open */
-  var file = root.device.open(root.rest, f);
+
+  // Node is weird and doesn't like opening NUL on Windows
+  // if its fs.openSync("NUL", "rs+") call, so we have to
+  // special case it and use \\\\.\\nul instead.
+  if (globalThis.process.platform === "win32" && path == "NUL") {
```
Comment (kopecs) [inline js/shared/core.js]:
```suggestion
  if (globalThis.process.platform === "win32" && path === "NUL") {
```

File: js/language_server/src/semgrep-lsp-bindings.ts [hunk 5]
```diff
@@ -1,3 +1,13 @@
+
+
+// If we're on Windows, we need to set the HOME and TEMP environment variables
+// to something that exists, otherwise tmp stuff and settings stuff will fail.
+if (process.platform === "win32") {
+  const os = require("os");
+  process.env["HOME"] = os.homedir();
```
Comment (kopecs) [inline js/language_server/src/semgrep-lsp-bindings.ts]: Would it be better to factor out usages of `$HOME` and `$TEMP` and have a platform function that reads, e.g., `HOMEDRIVE` and `HOMEPATH` if we're on windows instead? Unclear to me if the things that fail are our functions or stdlib.

**Review by kopecs — COMMENTED** [review]
Mostly just want a comment on the header path. Otherwise lgtm

File: js/shared/core.js (hunk 3 above)
Comment (ajbt200128) [reply js/shared/core.js:198]: ignoring this lmao

File: js/language_server/src/semgrep-lsp-bindings.ts (hunk 5 above)
Comment (ajbt200128) [reply js/language_server/src/semgrep-lsp-bindings.ts]: Fixed

File: libs/commons2/common2.ml [hunk 6]
```diff
@@ -2435,7 +2435,7 @@ let cat_excerpts file lines =
       let lines = List.sort compare lines in
       let rec aux acc lines count =
         let b, l =
-          try (true, input_line chan) with
+          try (true, Common.input_text_line chan) with
           | End_of_file -> (false, "")
```
Comment (mjambon) [inline libs/commons2/common2.ml:2439]: This should work.

**Review by mjambon — APPROVED** [review]
The use of `input_text_line` makes sense to me!

**Pull request merged**